# 期末程式作業I

姓名：涂育賢  
系級：經濟四  
學號：105208077  

## 使用bert來觀察句向量


以下模型使用方式可以參考此網址：https://github.com/hanxiao/bert-as-service  
這裡使用bert serving client作為線上伺服器來使用bert模型  
模型經過fine-tune (使用整理過的資料集)

### 載入套件

In [1]:
# 載入Bert線上伺服器
from bert_serving.client import BertClient
# 簡體繁體轉換工具
from opencc import OpenCC

# PCA套件
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# 使用bert-as-service線上伺服器服務
bc = BertClient()

### 輸入想觀察的句子

In [6]:
sentence = ['一個假裝快樂',
            '一個真心難過']

In [7]:
result = bc.encode(sentence)

### 句向量

In [8]:
result[0]

array([ 1.61591446e+00, -4.23978448e-01,  1.01122081e+00,  1.81399077e-01,
        1.78563982e-01, -2.90165693e-01,  2.38742113e-01, -1.78557113e-01,
        1.40623674e-01,  3.02694738e-01, -7.82200456e-01,  4.12410885e-01,
       -5.77114999e-01, -3.37580115e-01,  4.14301485e-01, -4.74271253e-02,
        9.14279580e-01,  6.43785894e-01, -4.49499995e-01,  7.74978936e-01,
        6.63304806e-01,  9.28839862e-01, -6.35536015e-01,  9.97821450e-01,
       -1.96804538e-01,  5.60819328e-01, -2.04970196e-01,  1.63485110e-01,
        5.49842678e-02, -4.30972040e-01, -1.23222268e+00,  4.17806625e-01,
       -7.15285182e-01,  1.61939964e-01,  2.12681651e-01,  6.91074193e-01,
       -4.78623867e-01, -8.01699877e-01,  2.56526709e-01, -3.98415029e-02,
       -6.16457880e-01, -7.14786589e-01,  2.63552666e-01,  3.18051800e-02,
       -9.37856808e-02, -8.30852211e-01, -2.06934720e-01,  1.81671474e-02,
        2.71195680e-01,  1.29049158e+00, -1.14301011e-01,  8.09836864e+00,
       -5.49369454e-01,  

In [9]:
result[1]

array([ 8.21436107e-01,  2.53808778e-02,  5.49332857e-01, -3.79476488e-01,
        2.66011596e-01, -2.98304826e-01,  6.07673287e-01, -2.66922206e-01,
       -6.32156134e-02,  6.15173101e-01, -1.25198305e+00,  1.38897285e-01,
       -2.92984664e-01, -7.90325105e-01,  8.07257593e-01, -7.35641122e-01,
        1.01556456e+00,  6.32343888e-01, -1.86265379e-01,  1.54849723e-01,
        7.44760185e-02,  6.50839686e-01, -6.10435069e-01,  8.86595666e-01,
        4.16118130e-02,  4.52538341e-01, -3.10780138e-01,  4.23418403e-01,
       -6.31748512e-02,  3.98212969e-02, -9.24482226e-01,  4.10271913e-01,
        3.29791486e-01,  7.01891541e-01, -1.06560156e-01,  8.90998244e-01,
       -5.53876348e-02, -9.01080310e-01, -1.86107010e-01, -1.99221581e-01,
       -5.82463503e-01, -1.85518503e-01,  8.57825994e-01,  1.32698849e-01,
        3.67630243e-01, -9.71277654e-01, -3.76742125e-01, -2.55494937e-02,
        3.14784974e-01,  3.87446731e-01, -1.91044703e-01,  7.92454338e+00,
        4.11122501e-01,  

### 透過cosine similarity來簡單判斷兩句話是否在向量空間中有一定的關係

In [10]:
from sklearn.metrics.pairwise import cosine_similarity

### 若在丟入cosine similarity前沒有進行reshpae會報錯

In [11]:
cosine_result = cosine_similarity(result[0],result[1])

ValueError: Expected 2D array, got 1D array instead:
array=[ 1.61591446e+00 -4.23978448e-01  1.01122081e+00  1.81399077e-01
  1.78563982e-01 -2.90165693e-01  2.38742113e-01 -1.78557113e-01
  1.40623674e-01  3.02694738e-01 -7.82200456e-01  4.12410885e-01
 -5.77114999e-01 -3.37580115e-01  4.14301485e-01 -4.74271253e-02
  9.14279580e-01  6.43785894e-01 -4.49499995e-01  7.74978936e-01
  6.63304806e-01  9.28839862e-01 -6.35536015e-01  9.97821450e-01
 -1.96804538e-01  5.60819328e-01 -2.04970196e-01  1.63485110e-01
  5.49842678e-02 -4.30972040e-01 -1.23222268e+00  4.17806625e-01
 -7.15285182e-01  1.61939964e-01  2.12681651e-01  6.91074193e-01
 -4.78623867e-01 -8.01699877e-01  2.56526709e-01 -3.98415029e-02
 -6.16457880e-01 -7.14786589e-01  2.63552666e-01  3.18051800e-02
 -9.37856808e-02 -8.30852211e-01 -2.06934720e-01  1.81671474e-02
  2.71195680e-01  1.29049158e+00 -1.14301011e-01  8.09836864e+00
 -5.49369454e-01  2.06965789e-01 -4.29772288e-01 -1.11046240e-01
  1.12200081e+00 -7.75914192e-02  5.66351593e-01  2.28669316e-01
 -1.43851191e-01 -1.86613783e-01  1.62831640e+00  3.81411761e-01
  2.84310873e-03  1.16230357e+00 -7.05412328e-01  5.02649486e-01
 -1.21809316e+00 -1.29359946e-01  8.42486918e-02  3.70992988e-01
 -1.86807930e-01  1.40878499e-01  3.05749625e-01  1.21856427e+00
 -6.60260797e-01  9.96433794e-01 -9.21881735e-01  3.00165683e-01
  9.93725598e-01 -2.20267311e-01  4.33758527e-01 -2.53740519e-01
  2.69474179e-01  3.37573647e-01 -9.01353538e-01  1.76664814e-01
  3.52785960e-02  3.86532485e-01 -8.08926105e-01  1.83625087e-01
 -3.19517702e-01  6.40401840e-01 -6.14785433e-01 -1.05561711e-01
  9.16807652e-01 -5.48496485e-01  7.56695390e-01  1.10704291e+00
  3.06322753e-01 -4.50320601e-01 -4.56838369e-01  8.21885884e-01
  1.33217144e+00  2.04700279e+00 -9.95577313e-04  5.13352692e-01
  1.12904596e+00 -9.83717263e-01  1.72428459e-01 -1.21329293e-01
 -4.12598908e-01  4.75869745e-01  1.80193171e-01 -5.68299070e-02
  5.24346292e-01 -6.30476654e-01 -7.53979981e-01 -2.04902068e-01
 -9.41717446e-01  7.93103650e-02 -1.53159723e-01 -7.26511955e-01
 -2.40366548e-01 -7.02352285e-01  1.51488990e-01  5.43261588e-01
 -5.11361837e-01 -1.85045511e-01 -5.59950650e-01 -4.37676042e-01
 -2.95091212e-01 -4.07613069e-01  1.84535041e-01 -7.54859269e-01
  6.87955678e-01 -6.42734289e-01 -2.94610918e-01  1.72277130e-02
 -1.03631008e+00  4.83506411e-01 -1.88075021e-01  4.32957888e-01
  7.20839322e-01  4.90056038e-01 -1.21729225e-01 -2.27826789e-01
  7.05283940e-01 -8.57487261e-01  3.69941622e-01 -4.61980462e-01
  5.45801222e-01  2.01531947e-01  6.90516829e-01  2.68057644e-01
 -9.39552188e-01  5.26807547e-01  5.74559569e-01  2.11679563e-01
 -9.32472050e-01  6.54439926e-01  2.15281650e-01  6.88932717e-01
  8.69894326e-01 -1.33912754e+00  5.59338152e-01  3.05429012e-01
 -4.43300486e-01 -1.11996603e+00  6.10946059e-01 -2.94841439e-01
  6.90532207e-01 -1.21855891e+00  1.28449768e-01  2.83686996e-01
 -9.24196541e-01 -4.59854275e-01  1.96045145e-01 -4.65415806e-01
 -1.23310268e+00 -3.06993902e-01 -1.15185249e+00 -9.45443273e-01
 -7.33912766e-01  3.54649335e-01 -6.15292192e-01 -4.76213008e-01
  7.04439998e-01 -3.05323184e-01  3.90004486e-01  3.04724634e-01
 -9.13262144e-02  4.57261741e-01 -2.16831446e-01 -6.10812604e-01
 -5.21291673e-01 -2.47662678e-01  5.25025129e-01  2.72403181e-01
  8.11094046e-02  7.27654576e-01 -5.52682221e-01  1.25551656e-01
  2.74985999e-01 -3.04954201e-01 -5.18159866e-01 -2.86006868e-01
  7.11215138e-01 -1.32459629e+00  1.42888165e+00  4.96424735e-01
 -2.51528859e-01 -1.73077695e-02 -4.46466684e-01  4.56474051e-02
 -2.73777127e-01 -6.46261454e-01 -4.02757317e-01  1.01911783e+00
 -1.38574407e-01 -6.73520625e-01  6.91937745e-01 -6.64953887e-02
  3.10967356e-01 -9.31131780e-01  7.18571365e-01 -6.09917700e-01
 -2.01894432e-01 -2.22189337e-01  6.36742786e-02 -1.68003455e-01
 -5.44193685e-01  5.64193785e-01  8.96621406e-01  5.45238554e-01
  4.90495153e-02  3.90228033e-01  5.82051456e-01  4.23721522e-01
 -2.84228593e-01 -1.99816242e-01  7.99525827e-02  5.43253839e-01
  7.71227032e-02  1.67021108e+00 -7.47322023e-01 -8.44418779e-02
 -1.35046273e-01  5.74648201e-01  8.16507876e-01 -7.98673272e-01
 -1.79195806e-01 -5.48169017e-01  1.75734252e-01  9.84582186e-01
 -4.41039771e-01 -1.90637514e-01  8.91847908e-01 -4.98557657e-01
  4.13400978e-01  1.76538453e-01  8.18198547e-02  1.38702169e-01
 -2.80837625e-01  1.08863151e+00 -9.89546999e-02 -1.38967931e+00
  7.52125859e-01 -7.93966830e-01 -5.66139758e-01 -2.92376149e-02
 -1.93475895e-02  9.12440121e-02 -9.23541307e-01  5.30541301e-01
 -1.99270338e-01 -3.44006121e-01  9.97669935e-01 -1.85022965e-01
 -3.83203804e-01 -4.79870468e-01  3.10938776e-01  2.87595779e-01
 -3.83470118e-01 -1.84507579e-01  2.66628087e-01  7.30603218e-01
  9.79999721e-01  9.63989615e-01  3.90871495e-01 -5.11810184e-01
  5.61824381e-01 -9.94164348e-01  2.50896811e+00 -1.33396033e-02
 -8.89267176e-02 -5.35466485e-02 -6.79073393e-01 -2.37480238e-01
 -1.45939123e-02  7.15535164e-01  6.93930328e-01  3.57214838e-01
 -1.19344883e-01 -3.68767768e-01  7.45602965e-01  4.69081551e-02
 -1.14653155e-01 -1.06996000e+00 -9.14433300e-01 -7.30950385e-02
  1.94780499e-01  2.05857262e-01 -1.05298430e-01  1.32901251e+00
 -5.03091253e-02  1.07223913e-02 -3.52938861e-01  4.93109614e-01
 -8.51542234e-01 -9.74389315e-01  4.48498458e-01  5.39409876e-01
  7.97222018e-01  8.82264555e-01  3.18648577e-01  4.54381220e-02
 -4.19152468e-01  5.17100692e-01 -1.23482630e-01  1.36772707e-01
  2.23640017e-02 -2.39937142e-01 -3.63963485e-01  8.66419435e-01
  1.34020969e-01  3.19156617e-01  6.60682321e-01  1.81126952e-01
  3.38613451e-01  2.67773718e-01 -7.32336342e-02 -9.43491384e-02
 -7.41701066e-01  2.37592250e-01  1.61382663e+00  6.55628294e-02
 -3.20327193e-01  7.31861949e-01 -3.85378957e-01 -1.20279506e-01
  3.80589545e-01 -4.56337959e-01  6.12167358e-01  1.80193260e-01
 -1.53175503e-01 -7.89841890e-01  2.99812872e-02 -3.51666883e-02
  9.91546512e-01 -2.56221890e-01 -1.07339181e-01  1.16910115e-02
 -3.48283738e-01 -1.84744954e-01 -2.05718413e-01 -4.41757500e-01
  6.56321406e-01 -6.00701988e-01  5.36842756e-02 -4.87195283e-01
 -8.61855984e-01  4.15402889e-01 -4.87661988e-01 -2.51767695e-01
 -5.80177248e-01  2.41217554e-01  7.98880696e-01 -7.57402420e-01
  9.15314257e-01  2.88569778e-01 -1.16267838e-01 -1.33485436e-01
 -2.41663247e-01  8.77367795e-01 -5.35385072e-01 -3.45851630e-02
 -7.77736425e-01 -7.12685883e-01 -3.29246998e-01 -5.80534697e-01
 -4.64061201e-01  5.97863346e-02  7.78162032e-02  1.06049515e-01
  3.75410497e-01 -8.92885804e-01  8.37401927e-01  5.86662889e-02
  2.53873199e-01 -7.82291889e-02 -5.98397017e-01 -7.32633099e-02
  5.27596235e-01 -1.31756866e+00 -3.14722359e-01  2.54337549e-01
 -1.02937365e+00  1.16179995e-02 -5.63109994e-01 -2.59246141e-01
  6.75388515e-01  8.77872169e-01  9.45701361e-01 -4.24208879e-01
  3.75905931e-01 -8.66356850e-01  4.64419350e-02 -9.75709110e-02
 -2.67888516e-01  1.00347567e+00  1.64666259e+00  4.74498980e-02
 -2.46334732e-01 -4.93275553e-01  5.22632480e-01  1.62442446e+00
 -2.97136873e-01 -2.08013102e-01  2.91881949e-01 -1.04723597e+00
 -5.83728969e-01 -5.76105416e-01  2.43529066e-01 -1.50329694e-01
 -8.93844008e-01 -1.01497269e+00  8.42102710e-03  1.44433994e-02
 -5.41033208e-01  1.14355080e-01  8.14642370e-01  4.74552542e-01
 -2.99427599e-01  4.48607147e-01  4.18413252e-01 -2.58816332e-01
 -1.11568063e-01  2.14807540e-02  3.61157686e-01 -1.09258032e+00
 -8.63639295e-01 -6.79162085e-01  4.75786507e-01  7.65030384e-02
 -1.23403084e+00  3.58032495e-01  1.36202037e-01  8.47843289e-01
 -2.09536850e-02  4.96496946e-01  5.55464804e-01 -2.74004757e-01
  5.16801655e-01  6.84962630e-01  1.71603188e-01 -1.33915412e+00
  1.42599687e-01 -8.71481001e-02 -3.16901743e-01 -1.83912233e-01
 -3.61663282e-01 -4.67409641e-01 -5.60482621e-01  7.12981150e-02
 -3.48300636e-01 -8.57894242e-01 -8.90209377e-01  1.43090114e-01
  1.70517278e+00  6.16746902e-01 -5.59343815e-01 -9.71056893e-02
 -3.72218758e-01 -5.40412739e-02  6.59207702e-01  3.68443616e-02
  2.19847858e-02 -8.87305498e-01 -2.97064841e-01 -1.48683679e+00
  7.27721930e-01  4.11358401e-02 -3.00349798e-02 -1.54541388e-01
 -8.23711991e-01 -4.16457742e-01  6.49441659e-01 -7.13264644e-02
 -3.27109963e-01 -1.29565150e-01  3.20935339e-01 -7.54356623e-01
 -1.00418627e+00 -5.39332449e-01  1.79895416e-01  1.19809520e+00
  8.02045092e-02  3.00153792e-01  2.41159603e-01 -3.68705839e-01
  5.52727461e-01 -1.10415363e+00  3.18060905e-01  7.06782401e-01
  2.39707515e-01 -7.14494660e-03  4.21682000e-01 -2.44536862e-01
  6.59995526e-02 -1.50020479e-03  7.73618102e-01 -6.84273005e-01
 -1.79845631e-01  2.28844479e-01  3.71848404e-01 -5.07606491e-02
 -1.47227481e-01 -1.31470054e-01 -3.96594182e-02 -1.90701395e-01
  4.73783225e-01  5.54991007e-01 -4.40661073e-01 -9.92420018e-02
 -8.07443976e-01 -1.11613142e+00 -8.53257701e-02  1.21706426e+00
  1.81420237e-01 -6.22406840e-01  3.91985327e-01 -1.31234437e-01
  1.83477458e-02  3.03035289e-01  1.19999394e-01  1.96356222e-01
 -6.04024291e-01 -1.20862305e+00 -1.00526547e+00  1.03530931e+00
  9.81450140e-01 -7.56009892e-02 -3.75044227e-01 -2.79693693e-01
  2.41475359e-01 -6.31748259e-01  3.19639072e-02 -1.04338169e-01
 -7.04995036e-01  9.81622040e-02 -8.22100461e-01  8.31186920e-02
 -2.93190777e-01  1.19477773e+00 -5.28307319e-01  1.34257686e+00
 -5.83087683e-01  2.47260571e-01  2.59946823e-01  2.88616270e-01
 -7.02087224e-01  3.81844759e-01  3.79199386e-01 -3.11774641e-01
 -6.97375596e-01 -4.74530160e-01  3.09111744e-01 -1.52472183e-01
 -1.97450325e-01  9.13651347e-01 -2.03736514e-01  8.58173013e-01
  1.06306863e+00  2.87194639e-01 -3.87718499e-01  1.66582584e-01
  4.24901068e-01 -9.92138028e-01  6.60629272e-01  1.64784968e-01
 -4.58714545e-01 -3.69644254e-01 -4.66742277e-01  2.76790172e-01
 -2.12238118e-01 -3.58617842e-01  6.72158599e-01 -1.69441566e-01
  5.34874141e-01 -1.27180308e-01 -7.44043887e-02  5.72752535e-01
 -2.02896059e-01  1.01469599e-01  3.92881304e-01 -4.08466488e-01
  2.46869728e-01  4.67287488e-02  8.06682482e-02  1.71005046e+00
  1.20197855e-01 -1.79245308e-01  3.82484019e-01 -6.67816281e-01
 -9.18913400e-04 -2.15136781e-01 -3.92274231e-01  5.64199805e-01
  7.81868994e-01 -4.74093467e-01  7.84080267e-01  4.71089303e-01
 -6.64480209e-01  3.51225048e-01 -1.07306108e-01  4.47922051e-01
  8.97681355e-01  1.02324975e+00  3.99432033e-01  5.29011488e-01
 -3.26742351e-01  4.71121788e-01  1.14320195e+00 -1.03908622e+00
 -9.88057971e-01  2.17502937e-01  1.36801922e+00 -2.39648610e-01
  3.71026754e-01 -2.51281887e-01 -4.38783407e-01  9.30148363e-02
 -9.90284443e-01 -6.81005597e-01  3.04309219e-01 -1.32880479e-01
  2.24062651e-01  4.58371371e-01  3.36315691e-01  1.16917968e+00
  1.03017640e+00 -7.08679855e-01 -5.27558863e-01 -3.21902990e-01
 -3.92616481e-01  2.97005653e-01  3.46930742e-01  3.51633608e-01
 -1.62133053e-01 -5.13216734e-01  1.77203134e-01  5.86234212e-01
  9.60846722e-01  1.16311617e-01 -2.68587321e-02  2.51402825e-01
 -1.54020086e-01  7.47928202e-01  8.80773127e-01  1.72411859e-01
  4.42129046e-01  5.66017389e-01  2.86988765e-01 -5.17546892e-01
  3.48580420e-01 -5.14200330e-01 -4.01990831e-01 -2.86429256e-01
 -2.70502925e-01  5.08956075e-01 -7.55937696e-01 -6.14468396e-01
  5.76493025e-01  2.54062742e-01  3.27967852e-01  3.73330027e-01
  3.14803481e-01 -7.32110977e-01 -8.70283604e-01 -6.82545662e-01
  2.88609296e-01  8.38712275e-01 -7.88323164e-01 -8.92487526e+00
 -2.48536378e-01 -8.46529245e-01 -9.68738645e-02 -7.92683840e-01
  2.20285952e-01 -5.05044281e-01  8.61911595e-01  2.75162667e-01
  6.24721348e-01 -1.07909310e+00 -8.32393587e-01 -1.72281098e-02
  5.66677749e-01  2.74292946e-01 -8.58048737e-01 -9.02758718e-01
  1.23963249e+00 -2.61050642e-01  7.35376239e-01  5.41833580e-01
 -4.78622079e-01 -3.15838873e-01 -5.61526358e-01  5.39460182e-01
 -3.44174683e-01  3.89312096e-02  3.63841385e-01  9.64357793e-01
  3.18027496e-01  7.40537047e-02 -1.17079306e+00 -1.67502284e-01
 -1.14097498e-01 -8.81438982e-03 -6.62961245e-01 -3.31783056e-01
 -7.34027803e-01  3.60606879e-01 -9.65066813e-03 -7.28977844e-02
  2.95199603e-01  1.89853394e+00  2.27738038e-01  5.69274068e-01
 -2.58187890e-01  1.32202303e+00 -4.88835514e-01  8.54296029e-01
 -1.19840227e-01 -1.30999669e-01  1.31165624e+00 -4.38589543e-01
 -3.88589203e-02  3.28075707e-01  5.18366098e-01  3.66002023e-01
  2.56645083e-01 -2.43430391e-01  3.85848194e-01 -4.23495471e-01
  6.63433731e-01 -1.44379461e+00  3.64761144e-01  5.28196571e-03
 -4.71263856e-01 -1.27291933e-01  2.80187219e-01 -2.04273045e-01
 -8.31920564e-01  2.13650078e-01  4.65633661e-01 -8.67833853e-01].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

In [12]:
cosine_result = cosine_similarity(result[0].reshape(1,-1),result[1].reshape(1,-1))

In [24]:
print(cosine_result)

[[0.89639187]]


### 小結

句子一：一個假裝快樂  
句子二：一個真心難過  

cosine similarity：0.8964  
代表這兩句話在向量空間中的相似程度有0.896